### ISSUE TITLE
[To Improve] Conditional return Type datas

### ISSUE THREAD

**Issue by charles4201** [thecodingmachine/graphqlite#33]
We have a special case, for a ReservationType, we want to return the data only if it belongs to this person, so we have to add conditions on every "**fields**" today, as below. We would like to declare one condition for all or only certain "**fields**" without repeating functions in AbstractReservationType.

```
/**
     * @Field()
     * @param Reservation $reservation
     * @return Bill|null
     */
    public function getBill(Reservation $reservation): ?Bill
    {
        if ((($reservation->getClient() !== null) && ($this->userService->getUserId() === $reservation->getClient()->getId())) || $this->rightsService->isAllowed('IS_ADMIN')) {
            return $reservation->getBill();
        }
        return null;
    }

    /**
     * @Field()
     * @param Reservation $reservation
     * @return Client|null
     */
    public function getClient(Reservation $reservation): ?Client
    {
        if ((($reservation->getClient() !== null) && ($this->userService->getUserId() === $reservation->getClient()->getId())) || $this->rightsService->isAllowed('IS_ADMIN')) {
            return $reservation->getClient();
        }
        return null;
    }
```

**Comment by moufmouf** [comment]
Hey @charles4201 ,

This is something I'll be looking at in the next release (3.1), as this is quite a common need.
I'm not sure how we can solve it, but my guess is that we should probably work on custom annotations that can register a kind of middleware that applies on the Field.

Something like:

```php
    /**
     * @Field()
     * @MyCustomRight
     */
    public function getClient(Reservation $reservation): ?Client
```

and

```php
class MyCustomRightService implements GraphQLMiddleware
{
    public function __construct(RightService $rightService) { /*....*/ }

    public function process(object $obj, HandlerInterface $handler) : ?object
    {
        if ($obj instanceof Reservation) {
            if (!(($reservation->getClient() === null) || ($this->userService->getUserId() === $reservation->getClient()->getId())) && $this->rightsService->isAllowed('IS_ADMIN')) {
                return null;
            }
        }
        return $handler->handle($obj);
    }
}
```

**Comment by moufmouf** [comment]
This is actually now solved by the brand new @Security annotation #132 

See documentation: https://graphqlite.thecodingmachine.io/docs/next/fine-grained-security
